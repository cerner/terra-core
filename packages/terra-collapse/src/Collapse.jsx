import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Arrange from 'terra-arrange';
import Button from 'terra-button';
import IconChevronRight from 'terra-icon/lib/icon/IconChevronRight';
import Toggler from 'terra-toggler';
import 'terra-base/lib/baseStyles';
import './Collapse.scss';

const propTypes = {
  /**
   * Content in the body of the collapse component that will be expanded or collapsed
   */
  children: PropTypes.node.isRequired,
  /**
   * Sets the text inside of the button when the collapse component is closed
   */
  closedButtonText: PropTypes.string.isRequired,
  /**
   * Used to set props and HTML attributes on the collapse button
   */
  buttonAttrs: PropTypes.object,
  /**
   * Icon displayed next to text content within the collapse button
   */
  icon: PropTypes.element,
  /**
   * Sets the collapse to be animated when it is opened or closed
   */
  isAnimated: PropTypes.bool,
  /**
   * Used to turn on animation on collapse button icon
   */
  isIconAnimated: PropTypes.bool,
  /**
   * Sets the collapse to only display and icon. Uses closedButtonText prop as an aria-label on the button.
   */
  isIconOnly: PropTypes.bool,
  /**
   * Sets the collapse initial state to open
   */
  isInitiallyOpen: PropTypes.bool,
  /**
   * Callback function triggered when collapse is closed
   */
  onClose: PropTypes.func,
  /**
   * Callback function triggered when collapse is opened
   */
  onOpen: PropTypes.func,
  /**
   * Sets the text inside of the button when the collapse component is open
   */
  openedButtonText: PropTypes.string,
};

const defaultProps = {
  isAnimated: false,
  isIconAnimated: false,
  isIconOnly: false,
  isInitiallyOpen: false,
  icon: <IconChevronRight />,
};

class Collapse extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpen: props.isInitiallyOpen,
    };

    this.handleOnClick = this.handleOnClick.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.isInitiallyOpen === this.props.isInitiallyOpen) {
      return;
    }

    this.setState({ isOpen: nextProps.isInitiallyOpen });
  }

  handleOnClick(e) {
    e.preventDefault();

    // Fire event from toggler handlers
    if (!this.state.isOpen && this.props.onOpen) {
      this.props.onOpen();
    } else if (this.state.isOpen && this.props.onClose) {
      this.props.onClose();
    }

    this.setState(prevState => ({
      isOpen: !prevState.isOpen,
    }));
  }

  render() {
    const {
      buttonAttrs,
      children,
      closedButtonText,
      icon,
      isAnimated,
      isIconAnimated,
      isIconOnly,
      isInitiallyOpen,
      onClose,
      onOpen,
      openedButtonText,
      ...customProps
     } = this.props;
    // Set openHeaderText to the same value as closedHeaderText if its not already set
    const normalizedOpenButtonText = openedButtonText || closedButtonText;
    const buttonText = !this.state.isOpen ? closedButtonText : normalizedOpenButtonText;
    const collapseClass = classNames([
      'terra-Collapse',
      { 'is-open': this.state.isOpen },
      { 'is-icon-animated': isIconAnimated },
      customProps.className,
    ]);

    let collapseButton;
    if (isIconOnly) {
      collapseButton = (
        <Button
          {...buttonAttrs}
          aria-expanded={this.state.isOpen}
          aria-label={closedButtonText}
          onClick={this.handleOnClick}
        >
          <span className="terra-Collapse-icon">{icon}</span>
        </Button>
      );
    } else {
      collapseButton = (
        <Button
          {...buttonAttrs}
          aria-expanded={this.state.isOpen}
          onClick={this.handleOnClick}
        >
          <Arrange
            align="center"
            fitStart={<span className="terra-Collapse-icon">{icon}</span>}
            fill={<span className="terra-Collapse-buttonText">{buttonText}</span>}
          />
        </Button>
      );
    }

    return (
      <div {...customProps} className={collapseClass}>
        {collapseButton}
        <Toggler isOpen={this.state.isOpen} isAnimated={this.props.isAnimated}>
          {this.props.children}
        </Toggler>
      </div>
    );
  }
}

Collapse.propTypes = propTypes;
Collapse.defaultProps = defaultProps;

export default Collapse;
