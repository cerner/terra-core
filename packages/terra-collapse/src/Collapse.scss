@import '~terra-mixins';

.terra-Collapse {
  display: block;
}

.terra-Collapse-icon {
  display: inline-block;
  font-size: 0.8em;
  transform: rotate(0deg);
}

.terra-Collapse.is-open .terra-Collapse-icon {
  transform: rotate(90deg);
}

.terra-Collapse.is-icon-animated .terra-Collapse-icon {
  transition: transform 0.15s ease-in;
}

.terra-Collapse.is-open.is-icon-animated .terra-Collapse-icon {
  transition: transform 0.15s ease-in;
}

.terra-Collapse-buttonText {
  @include terra-padding-start(0.3rem);
}
