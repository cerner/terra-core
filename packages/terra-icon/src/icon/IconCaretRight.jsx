import React from 'react';
import classNames from 'classnames';
import IconBase from '../IconBase';

const SvgIcon = customProps => {
  const attributes = Object.assign({}, customProps);
  attributes.className = classNames("is-themeable", attributes.className);

  return (
    <IconBase {...attributes}>
      <path d="M12 0l24 24-24 24V0z" ></path>
    </IconBase>
  )
};

SvgIcon.displayName = "IconCaretRight";
SvgIcon.defaultProps = {"className":"is-themeable","viewBox":"0 0 48 48","xmlns":"http://www.w3.org/2000/svg","isBidi":true};

export default SvgIcon;
