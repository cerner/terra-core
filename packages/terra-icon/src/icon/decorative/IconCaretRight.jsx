/* eslint-disable */
import React from 'react';
import IconBaseDec from '../../IconBaseDec';

const SvgIcon = (customProps) => {
  const attributes = Object.assign({}, customProps);

  return (
    <IconBaseDec {...attributes}>
      <path d="M12 0l24 24-24 24V0z" ></path>
    </IconBaseDec>
  );
};

SvgIcon.displayName = "IconCaretRightDec";
SvgIcon.defaultProps = {"className":"","viewBox":"0 0 48 48","xmlns":"http://www.w3.org/2000/svg","isBidi":true};

export default SvgIcon;
/* eslint-enable */
