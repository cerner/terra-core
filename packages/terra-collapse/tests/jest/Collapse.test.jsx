import React from 'react';
import IconCaretRight from 'terra-icon/lib/icon/IconCaretRight';
import Collapse from '../../src/Collapse';


describe('Collapse', () => {
  // Snapshot Tests
  it('should render a default collapse', () => {
    const collapse = render(<Collapse closedButtonText="Show">Test</Collapse>);
    expect(collapse).toMatchSnapshot();
  });

  it('should render an animated collapse', () => {
    const collapse = render(<Collapse closedButtonText="Show" isAnimated>Test</Collapse>);
    expect(collapse).toMatchSnapshot();
  });

  it('should render an initially open collapse', () => {
    const collapse = render(<Collapse closedButtonText="Show" isInitiallyOpen>Test</Collapse>);
    expect(collapse).toMatchSnapshot();
  });

  it('should render an icon only collapse', () => {
    const collapse = render(<Collapse closedButtonText="Show" isIconOnly>Test</Collapse>);
    expect(collapse).toMatchSnapshot();
  });

  // Event Tests
  it('should render an open collapse when clicked', () => {
    const collapse = mount(<Collapse closedButtonText="Show">Test</Collapse>);
    collapse.find('button').simulate('click');
    expect(collapse).toMatchSnapshot();
  });

  // Prop Tests
  it('should set buttonAttrs prop correctly', () => {
    const collapse = shallow(<Collapse buttonAttrs={{ variant: 'link' }} closedButtonText="Show">Test</Collapse>);
    expect(collapse.unrendered.props.buttonAttrs).toEqual({ variant: 'link' });
  });

  it('should set children prop correctly', () => {
    const collapse = shallow(<Collapse closedButtonText="Show">Test</Collapse>);
    expect(collapse.unrendered.props.children).toEqual('Test');
  });

  it('should set closedButtonText prop correctly', () => {
    const collapse = shallow(<Collapse closedButtonText="Show">Test</Collapse>);
    expect(collapse.unrendered.props.closedButtonText).toEqual('Show');
  });

  it('should set icon prop correctly', () => {
    const collapse = shallow(<Collapse closedButtonText="Show">Test</Collapse>);
    expect(collapse).toMatchSnapshot();
  });

  it('should set icon prop correctly with custom icon', () => {
    const collapse = shallow(<Collapse closedButtonText="Show" icon={<IconCaretRight />}>Test</Collapse>);
    expect(collapse).toMatchSnapshot();
  });

  it('should set isAnimated prop correctly', () => {
    const collapse = shallow(<Collapse closedButtonText="Show" isAnimated>Test</Collapse>);
    expect(collapse.unrendered.props.isAnimated).toEqual(true);
  });

  it('should set isIconAnimated prop correctly', () => {
    const collapse = shallow(<Collapse closedButtonText="Show" isIconAnimated>Test</Collapse>);
    expect(collapse.unrendered.props.isIconAnimated).toEqual(true);
  });

  it('should set isIconOnly prop correctly', () => {
    const collapse = shallow(<Collapse closedButtonText="Show" isIconOnly>Test</Collapse>);
    expect(collapse.unrendered.props.isIconOnly).toEqual(true);
  });

  it('should set isInitiallyOpen prop correctly', () => {
    const collapse = shallow(<Collapse closedButtonText="Show" isInitiallyOpen>Test</Collapse>);
    expect(collapse.unrendered.props.isInitiallyOpen).toEqual(true);
  });

  it('should set onOpen prop correctly', () => {
    // eslint-disable-next-line no-alert
    const collapse = shallow(<Collapse closedButtonText="Show" onOpen={() => alert('Opened')}>Test</Collapse>);
    expect(collapse.unrendered.props.onOpen.toString()).toEqual('function onOpen() {return alert(\'Opened\');}');
  });

  it('should set onClose prop correctly', () => {
    // eslint-disable-next-line no-alert
    const collapse = shallow(<Collapse closedButtonText="Show" onClose={() => alert('Closed')}>Test</Collapse>);
    expect(collapse.unrendered.props.onClose.toString()).toEqual('function onClose() {return alert(\'Closed\');}');
  });

  it('should set openedButtonText prop correctly', () => {
    const collapse = shallow(<Collapse closedButtonText="Show" openedButtonText="Hide">Test</Collapse>);
    expect(collapse.unrendered.props.openedButtonText).toEqual('Hide');
  });

 // Attributes
  it('should merge classes passed in with attributes', () => {
    const collapse = shallow(<Collapse className="TestClass" closedButtonText="Show">Test</Collapse>);
    expect(collapse.prop('className')).toContain('TestClass');
  });

  it('should merge ids passed in with attributes', () => {
    const collapse = shallow(<Collapse id="TestId" closedButtonText="Show">Test</Collapse>);
    expect(collapse.prop('id')).toContain('TestId');
  });

  it('should append data passed in with attributes', () => {
    const collapse = shallow(<Collapse data-terra-text-mock="MockData" closedButtonText="Show">Test</Collapse>);
    expect(collapse.prop('data-terra-text-mock')).toContain('MockData');
  });

  it('should append styles passed in with attributes', () => {
    const collapse = shallow(<Collapse style={{ height: '100px' }} closedButtonText="Show">Test</Collapse>);
    expect(collapse.prop('style')).toEqual({ height: '100px' });
  });

  // Error Handling Test
  it('should throw error for required children', () => {
    try {
      shallow(<Collapse closedButtonText="Show">Test</Collapse>);
    } catch (e) {
      expect(e.message).toContain('The prop `children` is marked as required');
    }
  });

  it('should throw error for required closedButtonText', () => {
    try {
      shallow(<Collapse>Test</Collapse>);
    } catch (e) {
      expect(e.message).toContain('The prop `closedButtonText` is marked as required');
    }
  });
});
