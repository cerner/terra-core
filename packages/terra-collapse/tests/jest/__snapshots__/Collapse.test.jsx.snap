exports[`Collapse should render a default collapse 1`] = `
<div
  class="terra-Collapse">
  <button
    aria-disabled="false"
    aria-expanded="false"
    class="terra-Button terra-Button--default"
    type="button">
    <div
      class="terra-Arrange">
      <div
        class="terra-Arrange-fitStart terra-Arrange-fitStart--center">
        <span
          class="terra-Collapse-icon">
          <svg
            aria-hidden="true"
            class="terra-Icon is-bidi"
            data-name="Layer 1"
            focusable="false"
            height="1em"
            viewbox="0 0 48 48"
            width="1em"
            xmlns="http://www.w3.org/2000/svg">
            <path
              d="M37.7 24L14.2 48l-3.9-3.8L30 24 10.3 3.8 14.2 0z" />
          </svg>
        </span>
      </div>
      <div
        class="terra-Arrange-fill terra-Arrange-fill--center">
        <span
          class="terra-Collapse-buttonText">
          Show
        </span>
      </div>
      <div
        class="terra-Arrange-fitEnd terra-Arrange-fitEnd--center" />
    </div>
  </button>
  <div
    aria-hidden="true"
    class="terra-Toggler" />
</div>
`;

exports[`Collapse should render an animated collapse 1`] = `
<div
  class="terra-Collapse">
  <button
    aria-disabled="false"
    aria-expanded="false"
    class="terra-Button terra-Button--default"
    type="button">
    <div
      class="terra-Arrange">
      <div
        class="terra-Arrange-fitStart terra-Arrange-fitStart--center">
        <span
          class="terra-Collapse-icon">
          <svg
            aria-hidden="true"
            class="terra-Icon is-bidi"
            data-name="Layer 1"
            focusable="false"
            height="1em"
            viewbox="0 0 48 48"
            width="1em"
            xmlns="http://www.w3.org/2000/svg">
            <path
              d="M37.7 24L14.2 48l-3.9-3.8L30 24 10.3 3.8 14.2 0z" />
          </svg>
        </span>
      </div>
      <div
        class="terra-Arrange-fill terra-Arrange-fill--center">
        <span
          class="terra-Collapse-buttonText">
          Show
        </span>
      </div>
      <div
        class="terra-Arrange-fitEnd terra-Arrange-fitEnd--center" />
    </div>
  </button>
  <div
    aria-hidden="true"
    class="terra-Toggler">
    <div>
      <div
        aria-hidden="false"
        class="rah-static"
        style="height:0px;overflow:hidden;">
        <div>
          Test
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Collapse should render an icon only collapse 1`] = `
<div
  class="terra-Collapse">
  <button
    aria-disabled="false"
    aria-expanded="false"
    aria-label="Show"
    class="terra-Button terra-Button--default"
    type="button">
    <span
      class="terra-Collapse-icon">
      <svg
        aria-hidden="true"
        class="terra-Icon is-bidi"
        data-name="Layer 1"
        focusable="false"
        height="1em"
        viewbox="0 0 48 48"
        width="1em"
        xmlns="http://www.w3.org/2000/svg">
        <path
          d="M37.7 24L14.2 48l-3.9-3.8L30 24 10.3 3.8 14.2 0z" />
      </svg>
    </span>
  </button>
  <div
    aria-hidden="true"
    class="terra-Toggler" />
</div>
`;

exports[`Collapse should render an initially open collapse 1`] = `
<div
  class="terra-Collapse is-open">
  <button
    aria-disabled="false"
    aria-expanded="true"
    class="terra-Button terra-Button--default"
    type="button">
    <div
      class="terra-Arrange">
      <div
        class="terra-Arrange-fitStart terra-Arrange-fitStart--center">
        <span
          class="terra-Collapse-icon">
          <svg
            aria-hidden="true"
            class="terra-Icon is-bidi"
            data-name="Layer 1"
            focusable="false"
            height="1em"
            viewbox="0 0 48 48"
            width="1em"
            xmlns="http://www.w3.org/2000/svg">
            <path
              d="M37.7 24L14.2 48l-3.9-3.8L30 24 10.3 3.8 14.2 0z" />
          </svg>
        </span>
      </div>
      <div
        class="terra-Arrange-fill terra-Arrange-fill--center">
        <span
          class="terra-Collapse-buttonText">
          Show
        </span>
      </div>
      <div
        class="terra-Arrange-fitEnd terra-Arrange-fitEnd--center" />
    </div>
  </button>
  <div
    aria-hidden="false"
    class="terra-Toggler">
    Test
  </div>
</div>
`;

exports[`Collapse should render an open collapse when clicked 1`] = `
<Collapse
  closedButtonText="Show"
  icon={
    <IconChevronRight
      className=""
      data-name="Layer 1"
      isBidi={true}
      viewBox="0 0 48 48"
      xmlns="http://www.w3.org/2000/svg" />
  }
  isAnimated={false}
  isIconOnly={false}
  isInitiallyOpen={false}>
  <div
    className="terra-Collapse is-open">
    <Button
      aria-expanded={true}
      isBlock={false}
      isCompact={false}
      isDisabled={false}
      isReversed={false}
      onClick={[Function]}
      type="button"
      variant="default">
      <button
        aria-disabled={false}
        aria-expanded={true}
        className="terra-Button terra-Button--default"
        disabled={false}
        onClick={[Function]}
        type="button">
        <Arrange
          align="center"
          fill={
            <span
              className="terra-Collapse-buttonText">
              Show
            </span>
          }
          fitStart={
            <span
              className="terra-Collapse-icon">
              <IconChevronRight
                className=""
                data-name="Layer 1"
                isBidi={true}
                viewBox="0 0 48 48"
                xmlns="http://www.w3.org/2000/svg" />
            </span>
          }>
          <div
            className="terra-Arrange">
            <div
              className="terra-Arrange-fitStart terra-Arrange-fitStart--center">
              <span
                className="terra-Collapse-icon">
                <IconChevronRight
                  className=""
                  data-name="Layer 1"
                  isBidi={true}
                  viewBox="0 0 48 48"
                  xmlns="http://www.w3.org/2000/svg">
                  <IconBase
                    ariaLabel={null}
                    className=""
                    data-name="Layer 1"
                    focusable={false}
                    height="1em"
                    isBidi={true}
                    isSpin={false}
                    viewBox="0 0 48 48"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg">
                    <svg
                      aria-hidden="true"
                      className="terra-Icon is-bidi"
                      data-name="Layer 1"
                      focusable={false}
                      height="1em"
                      viewBox="0 0 48 48"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg">
                      <path
                        d="M37.7 24L14.2 48l-3.9-3.8L30 24 10.3 3.8 14.2 0z" />
                    </svg>
                  </IconBase>
                </IconChevronRight>
              </span>
            </div>
            <div
              className="terra-Arrange-fill terra-Arrange-fill--center">
              <span
                className="terra-Collapse-buttonText">
                Show
              </span>
            </div>
            <div
              className="terra-Arrange-fitEnd terra-Arrange-fitEnd--center" />
          </div>
        </Arrange>
      </button>
    </Button>
    <Toggler
      isAnimated={false}
      isOpen={true}>
      <div
        aria-hidden={false}
        className="terra-Toggler">
        Test
      </div>
    </Toggler>
  </div>
</Collapse>
`;

exports[`Collapse should set icon prop correctly 1`] = `
<div
  className="terra-Collapse">
  <Button
    aria-expanded={false}
    isBlock={false}
    isCompact={false}
    isDisabled={false}
    isReversed={false}
    onClick={[Function]}
    type="button"
    variant="default">
    <Arrange
      align="center"
      fill={
        <span
          className="terra-Collapse-buttonText">
          Show
        </span>
      }
      fitStart={
        <span
          className="terra-Collapse-icon">
          <IconChevronRight
            className=""
            data-name="Layer 1"
            isBidi={true}
            viewBox="0 0 48 48"
            xmlns="http://www.w3.org/2000/svg" />
        </span>
      } />
  </Button>
  <Toggler
    isAnimated={false}
    isOpen={false}>
    Test
  </Toggler>
</div>
`;

exports[`Collapse should set icon prop correctly with custom icon 1`] = `
<div
  className="terra-Collapse">
  <Button
    aria-expanded={false}
    isBlock={false}
    isCompact={false}
    isDisabled={false}
    isReversed={false}
    onClick={[Function]}
    type="button"
    variant="default">
    <Arrange
      align="center"
      fill={
        <span
          className="terra-Collapse-buttonText">
          Show
        </span>
      }
      fitStart={
        <span
          className="terra-Collapse-icon">
          <IconCaretRight
            className=""
            isBidi={true}
            viewBox="0 0 48 48"
            xmlns="http://www.w3.org/2000/svg" />
        </span>
      } />
  </Button>
  <Toggler
    isAnimated={false}
    isOpen={false}>
    Test
  </Toggler>
</div>
`;
